oauth2_http/java/com/google/auth/oauth2/StsRequestHandler.java

```diff
@@ -0,0 +1,229 @@
+import com.google.api.client.http.UrlEncodedContent;
+import com.google.api.client.json.GenericJson;
+import com.google.api.client.json.JsonObjectParser;
+import com.google.api.client.json.JsonParser;
+import com.google.api.client.util.GenericData;
+import com.google.api.client.util.Joiner;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/** Implements the OAuth 2.0 token exchange based on https://tools.ietf.org/html/rfc8693. */
+public class StsRequestHandler {
+  private static final String TOKEN_EXCHANGE_GRANT_TYPE =
+      "urn:ietf:params:oauth:grant-type:token-exchange";
+  private static final String REQUESTED_TOKEN_TYPE =
+      "urn:ietf:params:oauth:token-type:access_token";
+  private static final String PARSE_ERROR_PREFIX = "Error parsing token response.";
+
+  private String tokenExchangeEndpoint;
+  private StsTokenExchangeRequest request;
+  private HttpRequestFactory httpRequestFactory;
+
+  @Nullable private HttpHeaders headers;
+  @Nullable private String internalOptions;
+
+  /**
+   * Internal constructor.
+   *
+   * @param tokenExchangeEndpoint the token exchange endpoint
+   * @param request the token exchange request
+   * @param headers optional additional headers to pass along the request
+   * @param internalOptions optional GCP specific STS options
+   * @return an StsTokenExchangeResponse instance if the request was successful
+   */
+  private StsRequestHandler(
+      String tokenExchangeEndpoint,
+      StsTokenExchangeRequest request,
+      HttpRequestFactory httpRequestFactory,
+      @Nullable HttpHeaders headers,
+      @Nullable String internalOptions) {
+    this.tokenExchangeEndpoint = tokenExchangeEndpoint;
+    this.request = request;
+    this.httpRequestFactory = httpRequestFactory;
+    this.headers = headers;
+    this.internalOptions = internalOptions;
+  }
+
+  public static Builder newBuilder(
+      String tokenExchangeEndpoint,
+      StsTokenExchangeRequest stsTokenExchangeRequest,
+      HttpRequestFactory httpRequestFactory) {
+    return new Builder(tokenExchangeEndpoint, stsTokenExchangeRequest, httpRequestFactory);
+  }
+
+  /** Exchanges the provided token for another type of token based on the RFC 8693 spec. */
+  public StsTokenExchangeResponse exchangeToken() throws IOException {
+    UrlEncodedContent content = new UrlEncodedContent(buildTokenRequest());
+
+    HttpRequest httpRequest =
+        httpRequestFactory.buildPostRequest(new GenericUrl(tokenExchangeEndpoint), content);
+    httpRequest.setParser(new JsonObjectParser(OAuth2Utils.JSON_FACTORY));
+    if (headers != null) {
+      httpRequest.setHeaders(headers);
+    }
+
+    try {
+      HttpResponse response = httpRequest.execute();
+      GenericData responseData = response.parseAs(GenericData.class);
+      return buildResponse(responseData);
+    } catch (IOException e) {
+      if (!(e instanceof HttpResponseException)) {
```

Comment:
instead of instanceof it's clearer to use two catch blocks here: first HttpResponseException, then IOException